libnetwork/diagnostic: clean up Server type

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Refactored `diagnostic.Server` to be more idiomatic.
- Un-embedded the mutex field to stop it from implementing `sync.Locker`
- Folded the `Init` method into the `diagnostic.New()` constructor because constructors are supposed to return constructed objects
- Refactored the diagnostic request handlers into standard `http.Handler` functions as there was no reason (aside from ignorance of the Go language) for the handlers to be implemented the way they were

**- How I did it**
By knowing what a closure is.

**- How to verify it**
Add a `network-diagnostic-port` entry to daemon.json, start the daemon and SIGHUP it to trigger a config reload. (The diagnostic server is not started until the config is reloaded. Smells like a bug...to address separately.) cURL the different endpoints to make sure they still work and don't crash the daemon. Initialize a swarm and check that the networkdb diagnostic endpoints are registered, work, and don't crash the daemon.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
libnetwork/diagnostic: un-embed sync.Mutex field

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libnetwork/diagnostic: drop Init method

Fold it into the constructor, because that's what the constructor is
supposed to do.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented on libnetwork/networkdb/networkdbdiagnostic.go:
```diff
@@ -271,22 +249,17 @@ func dbDeleteEntry(ctx interface{}, w http.ResponseWriter, r *http.Request) {
-		}
-		logger.Info("delete entry done")
-		diagnostic.HTTPReply(w, diagnostic.CommandSucceed(nil), json)
+	err := nDB.DeleteEntry(tname, nid, key)
+	if err != nil {
```
Micro-nit; this could be inlined;

```go
if err := nDB.DeleteEntry(tname, nid, key); err != nil {
```

corhere replied:
This comment gets hidden when I hide whitespace in the diff view, so I'm inclined to leave the change whitespace-only to keep the noise down for reviewers.

thaJeztah replied:
Yeah, for sure not a blocker in either way.

thaJeztah reviewed (approved):
LGTM

corhere commented on libnetwork/diagnostic/server.go:
```diff
@@ -49,17 +49,12 @@ type Server struct {
 // New creates a new diagnostic server
 func New() *Server {
-	return &Server{
+	s := &Server{
```
Not possible; there's an `s.RegisterHandler(...)` statement before `return s`.

thaJeztah replied:
OH! I left that comment on the first commit, and the second commit adds those 👍 

yup, all good (it was a nit either way)

corhere replied:
The commit you commented on:
https://github.com/moby/moby/blob/ef405764251bfa5ce4366e948618dabf046abf77/libnetwork/diagnostic/server.go#L50-L58

thaJeztah replied:
OH! You are absolutely right; I somehow misread the diff and clobbered it together with the code that was (re)moved.

thaJeztah commented:
some failures in CI, but not sure if related. I re-kicked CI, but in case failures are relevant, let me post them below;


docker-py:

```
tests/integration/api_network_test.py .......time="2023-11-01T15:27:40Z" level=error msg="error waiting for container: "
```

integration-cli:

```
=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestAPISwarmForceNewCluster (9.93s)
    docker_api_swarm_test.go:580: [d41c4cfe156b3] joining swarm manager [d0ba6fd1eec40]@0.0.0.0:2477, swarm listen addr 0.0.0.0:2478
    docker_api_swarm_test.go:595: assertion failed: error is not nil: error during connect: Post "http://%2Ftmp%2Fdocker-integration%2Fd0ba6fd1eec40.sock/v1.44/swarm/init": EOF: initializing swarm
    check_test.go:660: [d0ba6fd1eec40] daemon is not started
    check_test.go:661: [d0ba6fd1eec40] error removing /go/src/github.com/docker/docker/bundles/test-integration/TestDockerSwarmSuite/TestAPISwarmForceNewCluster/d0ba6fd1eec40/root/overlay2: unlinkat /go/src/github.com/docker/docker/bundles/test-integration/TestDockerSwarmSuite/TestAPISwarmForceNewCluster/d0ba6fd1eec40/root/overlay2/1caf9397bed147528b7693706d0960e404219302e63e5876353d1966603d01fb/merged: device or resource busy
    check_test.go:660: [d41c4cfe156b3] daemon is not started
    --- FAIL: TestDockerSwarmSuite/TestAPISwarmForceNewCluster (9.93s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite/TestSwarmInit (2.50s)
    docker_cli_swarm_test.go:124: assertion failed: 
        Command:  /usr/local/cli-integration/docker --host unix:///tmp/docker-integration/dc8b71f93c602.sock swarm init
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   error during connect: Post http://%2Ftmp%2Fdocker-integration%2Fdc8b71f93c602.sock/v1.30/swarm/init: EOF
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    check_test.go:660: [dc8b71f93c602] daemon is not started
    --- FAIL: TestDockerSwarmSuite/TestSwarmInit (2.50s)

=== FAIL: amd64.integration-cli TestDockerSwarmSuite (1521.38s)
```

thaJeztah commented:
same failure 3x in a row, so starting to look it may be related 😓

corhere commented:
```
panic: http: multiple registrations for /join
```
Yep, it's related. I did say that panicking when two handler patterns collide is a feature...

corhere committed:
libnetwork/diagnostic: use standard http.Handler

We don't need C-style callback functions which accept a void* context
parameter: Go has closures. Drop the unnecessary httpHandlerCustom type
and refactor the diagnostic server handler functions into closures which
capture whatever context they need implicitly.

If the node leaves and rejoins a swarm, the cluster agent and its
associated NetworkDB are discarded and replaced with new instances. Upon
rejoin, the agent registers its NetworkDB instance with the diagnostic
server. These handlers would all conflict with the handlers registered
by the previous NetworkDB instance. Attempting to register a second
handler on a http.ServeMux with the same pattern will panic, which the
diagnostic server would historically deal with by ignoring the duplicate
handler registration. Consequently, the first NetworkDB instance to be
registered would "stick" to the diagnostic server for the lifetime of
the process, even after it is replaced with another instance. Improve
duplicate-handler registration such that the most recently-registered
handler for a pattern is used for all subsequent requests.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented:
@corhere looks like it's fixed? 🎉

thaJeztah commented:
@akerouanton @robmry PTAL 🤗

robmry reviewed (approved):
LGTM - maybe a couple of extra tweaks ...

corhere committed:
libnetwork/diagnostic: lock mutex in help handler

Acquire the mutex in the help handler to synchronize access to the
handlers map. While a trivial issue---a panic in the request handler if
the node joins a swarm at just the right time, which would only result
in an HTTP 500 response---it is also a trivial race condition to fix.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah reviewed (approved):
still LGTM (nice catch on the typos!)


I'll take Rob's LGTM into account, and merge 👍
